### PR TITLE
Swarm Tests - WIP

### PR DESCRIPTION
**Pull request by VictorBjelkholm**
Not mergable yet

### CODE REVIEW

**Comment by VictorBjelkholm** [comment]
Depends on https://github.com/ipfs/js-ipfs/pull/960

**Comment by VictorBjelkholm** [comment]
Two things I hit during writing these tests:

* For some reason, we cannot listen to port 0 twice somehow... Having `/ip4/127.0.0.1/tcp/0` is fine if it's just one time, but if we do the same, I would expect us to get assigned two random ports from the OS, but it only happens the first time, then the second listener is stuck on port 0
* Calling `ipfs.swarm.peers` in the callback of `ipfs.swarm.connect` does not show the peer you just connected to. It seems there is some async task that happens in the background of `ipfs.swarm.connect` after the callback has been called, instead of waiting for the connection to be fully established.

**Comment by diasdavid** [comment]
> For some reason, we cannot listen to port 0 twice somehow... Having /ip4/127.0.0.1/tcp/0 is fine if it's just one time, but if we do the same, I would expect us to get assigned two random ports from the OS, but it only happens the first time, then the second listener is stuck on port 0

That is strange. Can you replicate the same behaviour using just TCP sockets in a simple Node.js script? 

> Calling ipfs.swarm.peers in the callback of ipfs.swarm.connect does not show the peer you just connected to. It seems there is some async task that happens in the background of ipfs.swarm.connect after the callback has been called, instead of waiting for the connection to be fully established.

This should not be the case, but it certainly happens in the 'dialing dance' inside libp2p-swarm. This work -- https://github.com/libp2p/js-libp2p-swarm/issues/24 -- would solve this. Can you make sure to add a test for this case (a failing one) commented out and pointing to that issue so that we know what's going on?

File: src/swarm.js [hunk 1]
```diff
@@ -107,6 +109,140 @@ module.exports = (common) => {
...
+              (cb) => {
+                nodeA.swarm.connect(nodeBAddress, (err) => {
+                  expect(err).to.not.exist()
+                  cb()
+                })
+              },
+              (cb) => {
+                // Waiting to make sure nodes are connected
+                setTimeout(cb, 1000)
+              },
```
Comment (diasdavid) [inline src/swarm.js]: Is this necessary for nodeA.swarm.peers or for nodeB.swarm.peers?

File: src/swarm.js [hunk 2]
```diff
@@ -107,6 +109,140 @@ module.exports = (common) => {
...
+          })
+          it('Connecting two peers with two addresses each', (done) => {
+            let nodeA
+            let nodeB
+            let nodeBAddress
+            const configA = getConfig([
+              // Need to explicitly declare the ports as 0 doesn't work with
+              // multiple swarm addresses apparently
+              // Not sure how to deal with the case if the ports are busy, this
+              // is introducing flakyness
```
Comment (diasdavid) [inline src/swarm.js]: This is really weird to me. See https://github.com/ipfs/interface-ipfs-core/pull/150#issuecomment-325125461

**Comment by VictorBjelkholm** [comment]
> That is strange. Can you replicate the same behaviour using just TCP sockets in a simple Node.js script?

Made some small test program but no problem there. Made a test-case for libp2p-swarm though that is currently failing, you can see it here: https://github.com/libp2p/js-libp2p-swarm/pull/227

File: src/swarm.js (hunk 2 above)
Comment (VictorBjelkholm) [reply src/swarm.js]: Test case to reproduce the issue: https://github.com/libp2p/js-libp2p-swarm/pull/227

Comment (diasdavid) [reply src/swarm.js]: Add a TODO here and create an issue to track this so that we can merge this PR as it is now.

**Comment by diasdavid** [comment]
@VictorBjelkholm can you confirm that:

- [ ] This was tested against js-ipfs-api
- [ ] https://github.com/ipfs/js-ipfs/pull/960#issuecomment-326524425 is all that is needed for js-ipfs?

**Comment by VictorBjelkholm** [comment]
@diasdavid yes to both of them

**Pull request merged**